lib/streamlit/elements/widgets/button_group.py

```diff
@@ -379,22 +484,59 @@ def _button_group(
         selection_visualization: ButtonGroupProto.SelectionVisualization.ValueType = (
             ButtonGroupProto.SelectionVisualization.ONLY_SELECTED
         ),
-        after_register_callback: Callable[
-            [RegisterWidgetResult[T]], RegisterWidgetResult[T]
-        ]
-        | None = None,
+        label: str | None = None,
+        label_visibility: LabelVisibility = "visible",
+        help: str | None = None,
     ) -> RegisterWidgetResult[T]:
+        _maybe_raise_selection_mode_warning(selection_mode)
+
+        parsed_selection_mode: ButtonGroupProto.ClickMode.ValueType = (
+            ButtonGroupProto.SINGLE_SELECT
+            if selection_mode == "single"
+            else ButtonGroupProto.MULTI_SELECT
+        )
+
+        # when selection mode is a single-value selection, the default must be a single
+        # value too.
+        if (
+            parsed_selection_mode == ButtonGroupProto.SINGLE_SELECT
+            and default is not None
+            and isinstance(default, Sequence)
+            and len(default) > 1
+        ):
+            raise StreamlitAPIException(
+                "The default argument to `st.button_group` must be a single value when "
+                "`selection_mode='single'`."
+            )
+
+        if style not in ["segment", "pills", "borderless"]:
+            raise StreamlitAPIException(
+                "The style argument must be one of ['segment', 'pills', 'borderless']. "
+                f"The argument passed was '{style}'."
+            )
+
         key = to_key(key)
 
-        check_widget_policies(self.dg, key, on_change, default_value=default)
+        _default = default
+        if default is not None and len(default) == 0:
+            _default = None
+
+        check_widget_policies(self.dg, key, on_change, default_value=_default)
+        if icons is not None and len(icons) != len(indexable_options):
+            raise StreamlitAPIException(
+                "The number of icons must match the number of options."
+            )
 
         widget_name = "button_group"
         ctx = get_script_run_ctx()
         form_id = current_form_id(self.dg)
         formatted_options = (
             indexable_options
             if format_func is None
-            else [format_func(option) for option in indexable_options]
+            else [
+                format_func(indexable_options[index], icons[index] if icons else None)
```

Comment:
Do we need to validate icons here?

Reply:
What do you mean with validation? If you need the length of the array, this happens further above already

Reply:
For all our `icon` parameters, we usually only set it as icon if `validate_icon_or_emoji` returns true. This method also contains some exceptions if it contains valid material icons: https://github.com/streamlit/streamlit/blob/671730333c6caebbe167cc25e383ca1263cbeb3f/lib/streamlit/string_util.py#L98

Reply:
https://github.com/streamlit/streamlit/blob/671730333c6caebbe167cc25e383ca1263cbeb3f/lib/streamlit/string_util.py#L60

Reply:
ahh good to know, thank you! I will add it 👍

Reply:
I have added it and also added a test for passing invalid values there